app/service/agent_svc.py

```diff
@@ -1,15 +1,21 @@
 import asyncio
 import json
 import typing
-from datetime import datetime
+from datetime import datetime, timedelta
 
 from app.service.base_service import BaseService
 
 
 class AgentService(BaseService):
 
-    def __init__(self):
+    def __init__(self, untrusted_timer):
         self.log = self.add_service('agent_svc', self)
+        self.loop = asyncio.get_event_loop()
+        self.untrusted_timer = untrusted_timer
+
+    async def start_sniffer_untrusted_agents(self):
```

Comment:
with the change to forcing untrusted_timer to be set, we can remove the if self.untrusted_timer... and we can combine the private function logic with this public function (add a docstring as well)

Reply:
by doing that - I think we can also consolidate this function a bit, so it looks like:
```
    async def start_sniffer_untrusted_agents(self):
        data_svc = self.get_service('data_svc')
        try:
            while True:
                await asyncio.sleep(self.untrusted_timer)
                trusted_agents = await data_svc.explode_agents(criteria=dict(trusted=1))
                for a in trusted_agents:
                    last_trusted_seen = datetime.strptime(a['last_trusted_seen'], '%Y-%m-%d %H:%M:%S')
                    silence_time = (datetime.now() - last_trusted_seen).total_seconds()
                    if silence_time > self.untrusted_timer:
                        await data_svc.update('core_agent', 'paw', a['paw'], data=dict(trusted=0))
                        self.log.debug('Agent %s is now untrusted' % a['paw'])
        except Exception:
            traceback.print_exc()
```

Reply:
I combined the logic of the private function with that of the public function, as you suggested.
Compared to your code, I left the part related to the calculation of the trust_time_left for each agent -- and the sleep with variable time depending on the various trust_time_left calculated.
I think it's important to adapt the sleep to these calculations. In fact, if we always sleep for 'self.untrusted_timer' seconds we may lose the discovery of untrusted agents.

Example, with 'untrusted_timer' = 200 secs:

1. at **time X**: the sniffer wakes up and finds no untrusted agent (but let's say that agent Z is silent for 180 secs).
2. at **time (X + 20)**: agent Z should be considered untrusted but the sniffer at most will detect it at time (x + 200).
3. at **time (X + 100)**: Agent Z sends a heartbeat or a result, and its 'last_trusted_seen' will be updated at this time because he is still considered trusted.
4. at **time (x + 200)**: the sniffer wakes up, but does not detect any unstrusted agents even now.

Reply:
agreed-- thanks for explaining.